### PR TITLE
Enable Pmemcheck on PPC64

### PR DESCRIPTION
**Pull request by shivaprasadbhat**
The patches here attempt add support for pmemcheck on PPC64 machines.

The first three commits  
459bce0        Powerpc, add support for the sync instruction with L = 2, 3 and 5.
95e6ae4        Powerpc, dcbf add check for L field and ISA version.
474f8c7         PowerPC: Fix the L field for the sync and dcbf instructions. 
are backports of the upstream commits available on tag VALGRIND_3_20_0.

The remaining 6 commits would be private to the pmem/valgrind tree. Out of those 6,
two commits just move around the existing tests to arch specific directories.

I have tested the changes using pmdk's unittests on
 - amd64: on laptop with qemu, virtual nvdimm
 - ppc64: POWER10 and POWER9 machines

also valgrind's regtest on amd64 shows no regressions on my setup and travis(internal).

Kindly give your feedback on the changes.

- - -
This change is [Reviewable](https://reviewable.io/reviews/pmem/valgrind/90)

### CODE REVIEW

**Review by lukaszstolarczuk — COMMENTED** [review]
@kilobyte, do you feel like testing this patch on your setup(s)...?

> Reviewed 44 of 53 files at r2.
> *[Reviewable](https://reviewable.io/reviews/pmem/valgrind/90#-:-NJp9BdC3oN3KMM-Z5P-:b-caun9)* status: 44 of 53 files reviewed, all discussions resolved

File: pmemcheck/pmc_replace_memcmp.c [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+   }
+   else if (loD < loS) {
+      return !(hiD < loS);
+   }
+   else {
+      /* They start at same place.  Since we know neither of them has
+         zero length, they must overlap. */
+      return True;
+   }
+}
```
Comment (kilobyte) [inline pmemcheck/pmc_replace_memcmp.c:150]: This logic can be simplified to `hiS<loD || hiD<loS`

**Comment by kilobyte** [comment]
I don't have any ppc hardware to test on, nor am I knowledgeable about the architecture.  
But as for a mere eyeball review, all changes look plausible.

I've spotted a single possible optimization but that's just a "if you're bored" thing.

Thus: LGTM.

**Comment by shivaprasadbhat** [comment]
> I've spotted a single possible optimization but that's just a "if you're bored" thing.

Thanks for the quick review. The optimisation would work for same length source and destination(true with current usage of memset|cmp|cpy), I might have to change is_overlap() to accept single length argument to avoid breaking it with differing lengths if someone uses it in future.

Let me know if you really want me to fix it.

> 
> Thus: LGTM.

Thank you.  Let me know if you need any more data points/test results etc. to get this merged.

**Comment by wlemkows** [comment]
Please rebase your PR to the new default branch pmem-3.20

**Comment by shivaprasadbhat** [comment]
> Please rebase your PR to the new default branch pmem-3.20

Done.

Thanks!

**Comment by lukaszstolarczuk** [comment]
Hi, thanks for the update.

As you stated in your original description:
> are backports of the upstream commits available on tag VALGRIND_3_20_0

perhaps now (on pmem-3.20), these commits are not required in this PR...? Either this or something else, but you have a conflict with the base branch, now...

**Comment by shivaprasadbhat** [comment]
> Hi, thanks for the update.
> 
> As you stated in your original description:
> 
> > are backports of the upstream commits available on tag VALGRIND_3_20_0
> 
> perhaps now (on pmem-3.20), these commits are not required in this PR...? Either this or something else, but you have a conflict with the base branch, now...

The 3.20 pulled in those 3 into the branch already. My pull request now is for the remaining 6 patches. 

I dont see a conflict on the pull request, as the number of commits now is 6, and it was 9 before. Do you see conflict? Am I missing something?

Thanks!

**Review by lukaszstolarczuk — COMMENTED** [review]
huh, sorry, I think it was my imagination 😄 something didn't refresh on my side, I guess.
Apologies!
And thanks again!

I'll just simply verify now if it works correctly with PMDK repository.

> *[Reviewable](https://reviewable.io/reviews/pmem/valgrind/90#-:-NLL-wu3CXH3_U17J5Dv:b-cqqpbt)* status: 40 of 53 files reviewed, all discussions resolved

**Comment by shivaprasadbhat** [comment]
> I'll just simply verify now if it works correctly with PMDK repository.

If verifying on PPC, some PMDK test specific fixes are required for few of those tests to pass. Let me know if you want me to push them to my PMDK tree for you to pick. They need some more cleanup though, so holding back.

Thanks!

**Comment by lukaszstolarczuk** [comment]
> > I'll just simply verify now if it works correctly with PMDK repository.
> 
> If verifying on PPC, some PMDK test specific fixes are required for few of those tests to pass. Let me know if you want me to push them to my PMDK tree for you to pick. They need some more cleanup though, so holding back.
> 
> Thanks!

Great!
If you have fixes ready for PMDK, then, please, go ahead and push them to the PMDK repo.

**Comment by shivaprasadbhat** [comment]
> > > I'll just simply verify now if it works correctly with PMDK repository.
> > 
> > 
> > If verifying on PPC, some PMDK test specific fixes are required for few of those tests to pass. Let me know if you want me to push them to my PMDK tree for you to pick. They need some more cleanup though, so holding back.
> > Thanks!
> 
> Great! If you have fixes ready for PMDK, then, please, go ahead and push them to the PMDK repo.

I just sent the pull request on PMDK tree with those fixes here https://github.com/pmem/pmdk/pull/5518

Thanks!

**Comment by shivaprasadbhat** [comment]
Thank you Weronika!

Regards,
Shivaprasad

**Pull request merged**